### PR TITLE
[TEVA-1533] AWS IAM STS Assume Role

### PR DESCRIPTION
**Pull request by nickromneydfe**
## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-1533

## Changes in this PR:

- Bump AWS Terraform provider to `3.18.0`
- Create 3 groups
-- administrators
-- billingmanagers
-- developers
- Create 4 roles:
-- administrator
-- auditor
-- readonly
-- secreteditor
- Allow groups to assume per-function roles
- Create custom policy for Parameter Store access
- Enforce MFA on assume role action

## Next steps:

- Define group memberships

### CODE REVIEW

File: terraform/common/variables.tf [hunk 1]
```diff
@@ -8,4 +8,12 @@ locals {
...
   secondary_zone_name                   = "teaching-jobs.service.gov.uk"
   service_name                          = "teaching-vacancies"
+  secrets                               = yamldecode(data.aws_ssm_parameter.common_secrets.value)
+  groups = toset([
+    "${local.service_name}-admin",
+    "${local.service_name}-billing",
+    "${local.service_name}-dev",
+    "${local.service_name}-prod",
+    "${local.service_name}-readonly"
+  ])
```
Comment (vigneshmsft) [inline terraform/common/variables.tf:18]:
```suggestion
  groups = toset([for k,v in local.secrets : k])
```
You could generate the keys like this(👆🏼) and don't have to edit terraform code if the yaml keys change.

File: terraform/common/variables.tf [hunk 2]
```diff
@@ -8,4 +8,12 @@ locals {
   primary_zone_name                     = "teaching-vacancies.service.gov.uk"
   secondary_zone_name                   = "teaching-jobs.service.gov.uk"
   service_name                          = "teaching-vacancies"
```
Comment (vigneshmsft) [inline terraform/common/variables.tf:10]: maybe also introduce locals for
```
arn:aws:iam::${data.aws_caller_identity.current.account_id}:user/&{aws:username}
arn:aws:iam::${data.aws_caller_identity.current.account_id}:mfa/&{aws:username}
```

File: terraform/common/variables.tf (hunk 1 above)
Comment (nickromneydfe) [reply terraform/common/variables.tf]: Thanks. That's a nice improvement. I've separated out group membership from group policy attachment (as I might want to define a group, even if there are no members). 
I'm now setting the policies for the groups in the locals too

File: terraform/common/variables.tf (hunk 2 above)
Comment (nickromneydfe) [reply terraform/common/variables.tf:10]: Stylistically, I don't mind `${data.aws_caller_identity.current.account_id}` and we use it in the S3 code too as a prefix (It's defined as a data source in the S3 code).

File: terraform/common/variables.tf (hunk 1 above)
Comment (nickromneydfe) [reply terraform/common/variables.tf]: Assigning users to groups will become the responsibility of #digital-tools-support

File: terraform/common/groups.tf [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+resource aws_iam_policy allow_assume_role_readonly {
+  name        = "allow_assume_role_readonly"
+  description = "Allow assuming readonly role"
+  policy      = data.aws_iam_policy_document.allow_assume_role_readonly.json
+}
+
+resource aws_iam_role_policy_attachment readonly_role_policies {
+  for_each = toset([
```
Comment (saliceti) [inline terraform/common/groups.tf:97]: Can we replace these with default ReadOnlyAccess?

File: terraform/common/groups.tf [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+resource aws_iam_policy allow_assume_role_secreteditor {
+  name        = "allow_assume_role_secreteditor"
+  description = "Allow assuming secreteditor role"
+  policy      = data.aws_iam_policy_document.allow_assume_role_secreteditor.json
+}
+
+data aws_iam_policy_document parameter_store {
+  statement {
+    actions = [
+      "ssm:DescribeParameters",
```
Comment (saliceti) [inline terraform/common/groups.tf:123]: We could limit damages by removing DeleteParameter and DeleteParameters. So we would only delete a parameter as admin.

File: terraform/common/groups.tf [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      "ssm:PutParameter",
+      "ssm:DeleteParameter",
+      "ssm:GetParameterHistory",
+      "ssm:GetParametersByPath",
+      "ssm:GetParameters",
+      "ssm:GetParameter",
+      "ssm:DeleteParameters"
+    ]
+    resources = [
+      "arn:aws:ssm:${var.region}:${data.aws_caller_identity.current.account_id}:parameter/${local.service_name}/dev/*",
```
Comment (saliceti) [inline terraform/common/groups.tf:133]: Why not `"arn:aws:ssm:${var.region}:${data.aws_caller_identity.current.account_id}:parameter/${local.service_name}/*"` ?

File: terraform/common/data.tf [hunk 6]
```diff
@@ -0,0 +1,11 @@
+data aws_iam_group administrators {
```
Comment (saliceti) [inline terraform/common/data.tf:1]: Can we create the groups via terraform as well?
Also the group names may look nicer starting with a capital letter.

File: README.md [hunk 7]
```diff
@@ -56,13 +56,35 @@ brew tap homebrew/cask
...
 
-For programmatic access, including application deployment using terraform, create an API key for yourself in the AWS IAM section.
-Then use the [aws cli](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) to configure the access key locally.
+For programmatic access, including application deployment using Terraform, create access keys for yourself under [My Security Credentials](https://console.aws.amazon.com/iam/home?region=eu-west-2#/security_credentials) - note these credentials for later use when configuring the AWS CLI.
+
+In the same section, choose `Assign MFA device` and set up an authenticator app as a Virtual MFA device. 
+
+Then install and configure the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html).
+
+[Install for Mac with](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-mac.html#cliv2-mac-install-cmd)
```
Comment (saliceti) [inline README.md:70]: On mac it's common to use Homebrew: `brew install awscli`

File: README.md [hunk 8]
```diff
@@ -78,13 +100,31 @@ aws_access_key_id=<AWS_ACCESS_KEY_ID>
...
+role_arn = arn:aws:iam::530003481352:role/readonly
+source_profile = default
+mfa_serial = arn:aws:iam::530003481352:mfa/YOURUSERNAME
+
+[profile secreteditor]
+role_arn = arn:aws:iam::530003481352:role/secreteditor
+source_profile = default
+mfa_serial = arn:aws:iam::530003481352:mfa/YOURUSERNAME
+```
+
```
Comment (saliceti) [inline README.md:120]: Shall we show how to use the profile?
- `aws -profile xxx <command>`
- `AWS_PROFILE=xxx <aws command>` (used for run_in_env)

File: terraform/common/groups.tf [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  group      = aws_iam_group.developers.name
+  policy_arn = aws_iam_policy.allow_assume_role_readonly.arn
+}
+
+resource aws_iam_group_policy_attachment permit_developers_group_assume_role_secreteditor {
+  group      = aws_iam_group.developers.name
+  policy_arn = aws_iam_policy.allow_assume_role_secreteditor.arn
+}
+
+# Group creation
```
Comment (saliceti) [inline terraform/common/groups.tf:186]: We can remove the comments

**Review by saliceti — APPROVED** [review]

**Pull request merged**